src/main/java/org/broadinstitute/hellbender/engine/spark/JoinReadsWithVariants.java

```diff
@@ -66,7 +66,15 @@
 
         JavaPairRDD<GATKRead, Variant> allPairs = pairReadsWithVariants(readsWShards, variantsWShards);
 
-        return allPairs.distinct().groupByKey();
+        return allPairs.aggregateByKey(new HashSet<>(), (vs, v) -> {
+            if (v != null) { // pairReadsWithVariants can produce null variant
+                ((HashSet<Variant>) vs).add(v);
```

Comment:
Louis & I cherry-picked this change into master recently (since it was needed by us for performance reasons), and the best we could manage was to get it down to one typecast instead of two:

        final JavaPairRDD<GATKRead, ? extends Iterable<Variant>> gatkReadHashSetJavaPairRDD = allPairs.aggregateByKey(new HashSet<>(), (Set<Variant> vs, Variant v) -> {
            if (v != null) { // pairReadsWithVariants can produce null variant
                vs.add(v);
            }
            return vs;
        }, (Set<Variant> vs1, Set<Variant> vs2) -> {
            vs1.addAll(vs2);
            return vs1;
        });

        @SuppressWarnings("unchecked")
        final JavaPairRDD<GATKRead, Iterable<Variant>> gatkReadIterableJavaPairRDD = (JavaPairRDD<GATKRead, Iterable<Variant>>)gatkReadHashSetJavaPairRDD;

Whether this is an improvement or not depends largely on personal taste I think :) Watch out for the merge conflict here when you rebase!